website/blog/2021-02-04-raptorx.md

```diff
@@ -0,0 +1,152 @@
+
+For example, suppose a user sends the following query, where `ds` is a partition column:
+
+```sql
+select sum(col) from T where ds between '2021-01-01' and '2021-01-03'
+```
+
+The partially computed sum for each of `2021-01-01`, `2021-01-02`, and `2021-01-03` partitions (or more precisely the corresponding files) will be cached on leaf workers forming a "fragment result". Suppose the user sends another query:
+
+```sql
+select sum(col) from T where ds between '2021-01-01' and '2021-01-05'
+```
+
+Then, the leaf worker will directly fetch the fragment result for `2021-01-01`, `2021-01-02`, and `2021-01-03` from cache and just compute the partial sum for `2021-01-04` and `2021-01-05`.
+
+Note that the fragment result is based on the leaf query fragment, which could be highly flexible as users can add or remove filters or projects. The above example shows we can easily handle filters with only partition columns. In order to avoid the cache miss caused by frequently changing non-partition column filters, we introduced partition statistics based pruning. Consider the following query, where time is a non-partition column:
+
+```sql
+select sum(col) from T
+where ds between '2021-01-01' and '2021-01-05'
+and time > now() - interval '3' day
+```
+
+Note that `now()` is a function that has values changing all the time. If a leaf worker caches the plan fragment based on `now()`’s absolute value, there is mostly no chance to have a cache hit. However, if predicate `time > now() - interval '3' day` is a "loose" condition that is going to be true for most of the partitions, we can strip out the predicate from the plan during scheduling time. For example, if today was `2021-01-04`, we know for partition `ds = 2021-01-04`, predicate `time > now() - interval '3' day` is always true.
+
+More generically, consider the following figure that contains a predicate and 3 partitions (`A`, `B`, `C`) with stats showing min and max. When the partition stats domain does not have any overlap with the predicate domain (e.g. partition `A`), we could directly prune this partition without sending splits to workers. If the partition stats is completely contained by the predicate domain (e.g. Partition `C`), then we don’t need this predicate because it would always hold true for this specific partition, and we could strip the predicate when doing plan comparison. For other partitions that have some overlapping with the predicate, we have to scan the partition with the given filter.
+
+
+
+![Predicate Pruning](/img/blog/2021-02-04-raptorx/frc.png)
+
+## File Descriptor and Footer Cache
+
+A Presto worker caches the file descriptors in memory to avoid long openFile calls to remote storage. Also, a worker caches common columnar file and stripe footers in memory. The current supported file formats are ORC, DWRF, and Parquet. The reason to cache such information in memory is due to the high hit rate of footers as they are the indexes to the data itself.
```

Comment:
```suggestion
A Presto worker caches the file descriptors in memory to avoid long `openFile` calls to remote storage. Also, a worker caches common columnar file and stripe footers in memory. The current supported file formats are ORC, DWRF, and Parquet. The reason to cache such information in memory is due to the high hit rate of footers as they are the indexes to the data itself.
```